File: packages/native-prover/ios/Sources/NativeProverPlugin/MidenNativeProverPlugin.swift
import Capacitor
import Foundation
import MidenMobileProver

// Capacitor plugin that proves a Miden transaction natively. Bridges JS
// calls (Uint8Array → base64 → bytes → C ABI → bytes → base64 → Uint8Array)
// onto a background queue so the WKWebView main thread stays responsive
// during the multi-second prove.
//
// Wire format matches `RemoteTransactionProver` / the web-sdk's
// `JsCallbackTransactionProver`: caller passes serialized `TransactionInputs`,
// receives serialized `ProvenTransaction`. Encoding is base64 because
// Capacitor's plugin bridge round-trips through JSON.
@objc(MidenNativeProverPlugin)
public class MidenNativeProverPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "MidenNativeProverPlugin"
    public let jsName = "MidenNativeProver"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "prove", returnType: CAPPluginReturnPromise)
    ]

    /// Single-flight prove queue. Concurrent calls would compete for the
    /// global rayon pool inside libmiden_mobile_prover.a, hurting per-prove
    /// latency for both. A serial queue gives predictable timings.
    private let proveQueue = DispatchQueue(label: "com.miden.native-prover.prove", qos: .userInitiated)

    public override func load() {
        super.load()
        NSLog("[MidenNativeProver] plugin loaded")
    }

    @objc func prove(_ call: CAPPluginCall) {
        NSLog("[MidenNativeProver] prove() called, callbackId=\(call.callbackId)")
        // Quick synchronous probe path so we can confirm the bridge returns
        // at all. Pass `probe: "echo"` to get an immediate resolve.
        if let probe = call.getString("probe") {
            NSLog("[MidenNativeProver] prove() probe='\(probe)', resolving synchronously")
            call.resolve(["probe": probe, "echo": "ok"])
            return
        }
        guard let inputBase64 = call.getString("input"),
              let inputData = Data(base64Encoded: inputBase64) else {
            NSLog("[MidenNativeProver] prove() rejected: bad input shape")
            call.reject("MidenNativeProver.prove requires `input` as a base64-encoded Uint8Array")
            return
        }
        NSLog("[MidenNativeProver] prove() dispatching native prove for \(inputData.count) input bytes")

        proveQueue.async { [weak self] in
            guard self != nil else { return }
            let t0 = Date()
            let result = Self.runProve(input: inputData)
            let elapsedMs = Date().timeIntervalSince(t0) * 1000.0
            NSLog("[MidenNativeProver] prove() native call returned in \(elapsedMs)ms")
            DispatchQueue.main.async {
                switch result {
                case .success(let output):
                    NSLog("[MidenNativeProver] prove() success, output=\(output.count) bytes")
                    call.resolve([
                        "output": output.base64EncodedString(),
                        "durationMs": elapsedMs
                    ])
                case .failure(let err):
                    NSLog("[MidenNativeProver] prove() failure: \(err)")
                    call.reject("MidenNativeProver: \(err)")
                }
            }
        }
    }

    private enum ProveError: Error, CustomStringConvertible {
        case badInput
        case proveFailed
        case unknownStatus(Int32)

        var description: String {
            switch self {
            case .badInput: return "input bytes did not decode as TransactionInputs"
            case .proveFailed: return "prover rejected the transaction"
            case .unknownStatus(let s): return "native prover returned unknown status \(s)"
            }
        }
    }

    private static func runProve(input: Data) -> Result<Data, ProveError> {
        // First attempt: pre-size the output buffer to 4 MB. Real proven
        // transactions on the devnet typical-claim path are well under
        // 1 MB; 4 MB gives plenty of headroom without forcing a retry.
        var cap = 4 * 1024 * 1024
        for _ in 0..<2 {
            var output = [UInt8](repeating: 0, count: cap)
            var written: Int = 0
            let status: Int32 = input.withUnsafeBytes { inputBytes -> Int32 in
                guard let inputBase = inputBytes.baseAddress?.assumingMemoryBound(to: UInt8.self) else {
                    return -1
                }
                return output.withUnsafeMutableBufferPointer { outBuf -> Int32 in
                    miden_prove_transaction(
                        inputBase,
                        input.count,
                        outBuf.baseAddress,
                        outBuf.count,
                        &written
                    )
                }
            }

            switch status {
            case 0:
                return .success(Data(output.prefix(written)))
            case -1:
                return .failure(.badInput)
            case -2:
                return .failure(.proveFailed)
            case -3:
                // BufferTooSmall — `written` is the required size. Grow
                // with slack and retry once. Two iterations is enough:
                // the FFI's `written` is exact.
                cap = written + 4096
                continue
            default:
                return .failure(.unknownStatus(status))
            }
        }
        return .failure(.unknownStatus(-99))
    }
}

File: packages/native-prover/android/src/main/java/com/miden/nativeprover/MidenNativeProverPlugin.kt
package com.miden.nativeprover

import android.util.Base64
import android.util.Log
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import java.util.concurrent.Executors

/**
 * Capacitor plugin that proves a Miden transaction natively. Bridges JS
 * calls (Uint8Array → base64 → bytes → JNI → bytes → base64 → Uint8Array)
 * onto a background executor so the WebView main thread stays responsive
 * during the multi-second prove.
 *
 * Mirrors the iOS plugin's shape (`MidenNativeProverPlugin.swift`). The
 * native code is the same Rust crate built once per target ABI; differences
 * between platforms are only at the bridge layer (Swift FFI on iOS, JNI
 * here).
 */
@CapacitorPlugin(name = "MidenNativeProver")
class MidenNativeProverPlugin : Plugin() {

    companion object {
        private const val TAG = "MidenNativeProver"

        init {
            // `libmiden_native_prover_jni.so` is the cdylib produced by
            // packages/native-prover/android/rust-bridge. It statically
            // links in the Rust prover (rayon + miden-tx/concurrent +
            // miden-crypto, the works) and exports the JNI symbol that
            // `external fun proveNative` below resolves against.
            System.loadLibrary("miden_native_prover_jni")
        }

        @JvmStatic
        private external fun proveNative(input: ByteArray): ByteArray
    }

    /**
     * Single-flight prove executor. Concurrent calls would compete for the
     * rayon global pool inside libmiden_native_prover_jni.so, hurting
     * per-prove latency for both. A single-thread executor gives
     * predictable timings — same intent as the iOS side's serial
     * DispatchQueue.
     */
    private val proveExecutor = Executors.newSingleThreadExecutor { r ->
        Thread(r, "miden-native-prover").apply { isDaemon = true }
    }

    override fun load() {
        super.load()
        Log.i(TAG, "plugin loaded")
    }

    @PluginMethod
    fun prove(call: PluginCall) {
        Log.i(TAG, "prove() called, callbackId=${call.callbackId}")

        // Sync probe path mirroring the iOS plugin — resolves immediately
        // so a caller can confirm the bridge round-trip works without
        // running a real prove.
        val probe = call.getString("probe")
        if (probe != null) {
            Log.i(TAG, "prove() probe='$probe', resolving synchronously")
            val out = JSObject()
            out.put("probe", probe)
            out.put("echo", "ok")
            call.resolve(out)
            return
        }

        val inputBase64 = call.getString("input")
        if (inputBase64 == null) {
            Log.w(TAG, "prove() rejected: missing `input`")
            call.reject("MidenNativeProver.prove requires `input` as a base64-encoded Uint8Array")
            return
        }

        val inputBytes: ByteArray = try {
            Base64.decode(inputBase64, Base64.DEFAULT)
        } catch (e: IllegalArgumentException) {
            Log.w(TAG, "prove() rejected: input not valid base64", e)
            call.reject("MidenNativeProver: input is not valid base64")
            return
        }

        Log.i(TAG, "prove() dispatching native prove for ${inputBytes.size} input bytes")

        proveExecutor.submit {
            val t0 = System.nanoTime()
            try {
                val outputBytes = proveNative(inputBytes)
                val elapsedMs = (System.nanoTime() - t0) / 1_000_000.0
                Log.i(TAG, "prove() success, output=${outputBytes.size} bytes in ${"%.1f".format(elapsedMs)}ms")
                val out = JSObject()
                out.put("output", Base64.encodeToString(outputBytes, Base64.NO_WRAP))
                out.put("durationMs", elapsedMs)
                call.resolve(out)
            } catch (e: Throwable) {
                Log.e(TAG, "prove() failure", e)
                call.reject("MidenNativeProver: ${e.message ?: e.javaClass.simpleName}")
            }
        }
    }
}
